### PR TITLE
Add Box Shadow support for featured image

### LINKED ISSUE #59067: Add Box Shadow support for featured image
## What problem does this address?
In templates, features image are a major design element for Archive pages, and it would be expanded the design tools considerably if the box shadow support that is now available for the image block could also be provided for the featured image block. 

## What is your proposed solution?
Open up box shadow support to featured image (among other blocks)

### PR DESCRIPTION
**Pull request by colinduwe**
## What?
This PR adds box shadow support to the Post Featured Image Block so it has the same controls as the core Image block. 

## Why?
This PR resolves https://github.com/WordPress/gutenberg/issues/59067 and brings more parity between the Image and Post Featured Image blocks.

## How?
This PR adds block support for box shadows. The shadow style needs to be applied to the image tag inside the figure tag like the border styles. The selectors have been specified so that global styles from theme.json and elsewhere are also applied to the image rather than the figure. 

In order for this to work, the shadow.php needed to be updated to support __experimentalSkipSerialization. Without it, the box shadow style is printed to the figure.  

NOTE: This PR would potentially affect any themes that were providing a shadow style definition to the core/post-featured-image. Currently, that style is applied to the figure. With this PR it is applied to the image. This allows the box shadow to work in conjunction with a border radius.

## Testing Instructions
1. Open a post or page
2. Add a featured image
3. Add a Featured Image Block
4. Add a box shadow style. It should be visible in the editor.
5. Adjust any of the other style controls. They should all work in harmony.
6. Test the result on the front end.

### Testing Instructions for Keyboard
1. Open a post or page
2. Add a featured image
3. Add a Featured Image Block
4. Navigate to the sidebar with the keyboard
5. Navigate to the Border and & Shadow section
6. Open the kabob menu to enable the shadow with the keyboard
7. Add a box shadow style. It should be visible in the editor.
8. Adjust any of the other style controls. They should all work in harmony.
9. Test the result on the front end.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/2152870/f833d4ff-56d3-4bbe-a3de-22aeb2bae4cc
(Note that I've added a shadow to my theme.json to demonstrate that users can override global styles)

### CODE REVIEW

**Comment by bph** [comment]
Thank you, @colinduwe, for tackling this issue with your PR! 
I tested on my local site, setting box shadow for Feature images on Archive Pages and for the Single post template. Both displayed the selected box shadow on the front end and in the editor. 

 @aaronrobertshaw would you have time to review on the PR?

File: lib/block-supports/shadow.php [hunk 1]
```diff
@@ -51,6 +51,10 @@ function gutenberg_apply_shadow_support( $block_type, $block_attributes ) {
 		return array();
 	}
 
+	if ( wp_should_skip_block_supports_serialization( $block_type, 'shadow' ) ) {
+		return array();
+	}
+
```
Comment (aaronrobertshaw) [inline lib/block-supports/shadow.php]: This bug fix regarding skipping serialization might be something we should split out into its own PR and backport to the WP 6.5 RC.

File: packages/block-library/src/post-featured-image/block.json [hunk 2]
```diff
@@ -86,6 +89,9 @@
 			"clientNavigation": true
 		}
 	},
+	"selectors": {
+		"shadow": ".wp-block-post-featured-image img, .wp-block-post-featured-image .wp-block-image__crop-area, .wp-block-post-featured-image .components-placeholder"
+	},
```
Comment (aaronrobertshaw) [inline packages/block-library/src/post-featured-image/block.json]: I don't think the Post Featured Image block has a cropper like the image block. So the cropper related selectors here are probably able to be dropped.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for putting this PR together @colinduwe, nice work! 👍 

I have a couple of quick thoughts and suggestions I've added as inline comments below. 

I'm also currently away at WordCamp Asia until later next week when I can take this for a proper spin and dig a bit deeper.

In the meantime, hopefully, the comments below help! 🤞

**Comment by colinduwe** [comment]
@aaronrobertshaw Updated the PR to removed the changes to the /supports/shadow.php. That's now a separate PR https://github.com/WordPress/gutenberg/pull/59887 and I removed the extraneous selector. 

Note that this PR will not work correctly without the other PR.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for iterating on this @colinduwe 🙇 

> Note that this PR will not work correctly without the other PR.

In cases like this it can be a good idea to base this PR's branch off the new PR's so it all works or update the PR description so the test instructions to highlight how to test it.

That said, it's trivial to cherry pick the other PR's commit, rebase, or checkout an earlier commit on this branch. So just something to keep in the back of your mind.

---

I've rebased this PR locally for testing after merging https://github.com/WordPress/gutenberg/pull/59887 and it works nicely for me. Great work! ✨ 

✅ Box shadow is correctly applied to inner img in the editor
✅ Box shadow is applied to inner img on the frontend
✅ Theme.json applied box shadows for featured images works
✅ Global Styles applied box shadows work for featured images
✅ Feature image placeholder displays box shadow in the editor

LGTM 🚢

**Comment by jasmussen** [comment]
Perhaps unintentionally, this PR changed how border and border-radius are applied to the Featured Image block. Before this PR, values were applied to the `img` element:

img element code

This meant you could apply a border and radius to the image, and it would properly affect the apeparance:

before the pr

This PR changed that so the styles are now applied to the wrapping figure element instead:

figure element code

That means the border radius doesn't affect the image inside at all:

after this pr

CC: @aaronrobertshaw @colinduwe @bph 

It's not clear to me whether there's an easy fix, hopefully we can just point the selector to the `img` again.

**Comment by colinduwe** [comment]
I can't seem to reproduce what you're showing. When I run Gutenberg/trunk locally the border and shadow styles are correctly applied to the image element as inline styles. That behavior is controlled in https://github.com/WordPress/gutenberg/blob/c50e0c408e7740c6a8a817c985831670d300f030/packages/block-library/src/post-featured-image/block.json#L68-L72
Because this PR adds 
https://github.com/WordPress/gutenberg/blob/c50e0c408e7740c6a8a817c985831670d300f030/packages/block-library/src/post-featured-image/block.json#L92-L94
It might be possible that it's interfering. The fix would perhaps be to move that selector stuff from inside the supports section for borders into the separate selectors section. 
@jasmussen since I can't replicate on my end, would you please try that and let me know if it changes anything?

On the front end, border and shadow styles should be removed from the figure element by "__experimentalSkipSerialization" and then added to the image element in index.php. Can you confirm that all's well on the front end?

**Comment by aaronrobertshaw** [comment]
There's a fix up for the border issue in https://github.com/WordPress/gutenberg/pull/60184. If you have time to review and test it @colinduwe, that would be great! 🙏 

Apologies for being a bit slow to note the fix here, while it was a simple tweak, I found [another issue](https://github.com/WordPress/gutenberg/pull/59295#issuecomment-2019196878) with the overlay being broken that isn't so straightforward.

**Comment by jasmussen** [comment]
Looks all solved. Thank you both!

**Comment by colinduwe** [comment]
Looks good to me as well. Thanks @aaronrobertshaw

**Comment by t-hamano** [comment]
I submitted #62207 because I noticed that the shadow was not being applied on block instances.
